File: samples/Test/Classes/MenuViewController.cpp
//
//  MenuViewController.cpp
//  Test
//
//  Created by renhongguang on 15/4/3.
//
//

#include "MenuViewController.h"
#include "MyWebViewController.h"
#include "FirstViewController.h"

#pragma execution_character_set("utf-8")

static const wchar_t* menuList[4] = { L"CrossApp", L"9秒" };

MenuViewController::MenuViewController()
{
	CANotificationCenter::getInstance()->addObserver(CALLBACK_BIND_1(MenuViewController::changeStatusBarOrientation, this), this, CAApplicationDidChangeStatusBarOrientationNotification);
}

MenuViewController::~MenuViewController()
{
    CANotificationCenter::getInstance()->removeObserver(this, CAApplicationDidChangeStatusBarOrientationNotification);
}

void MenuViewController::viewDidLoad()
{
    this->getView()->setColor(CAColor4B::CLEAR);
    
    DLayout tableViewLayout;
    DLayout logoLayout;
    const CAInterfaceOrientation& orientation = CAApplication::getApplication()->getStatusBarOrientation();
    if (orientation == CAInterfaceOrientation::Landscape)
    {
        tableViewLayout = DLayout(DHorizontalLayoutFill, DVerticalLayout_B_H(0, 400));
        logoLayout = DLayout(DHorizontalLayout_W_C(261, 0.5), DVerticalLayout_T_H(120, 258));
    }
    else
    {
        tableViewLayout = DLayout(DHorizontalLayoutFill, DVerticalLayout_T_B(450, 0));
        logoLayout = DLayout(DHorizontalLayout_W_C(261, 0.5), DVerticalLayout_T_H(120, 258));
    }
    
    m_pTableView = CATableView::createWithLayout(tableViewLayout);
    m_pTableView->setAllowsSelection(true);
    m_pTableView->setBackgroundImage(nullptr);
    m_pTableView->setSeparatorColor(ccc4(166, 166, 166,100));
    m_pTableView->setShowsScrollIndicators(false);
    m_pTableView->setScrollEnabled(false);
    this->getView()->addSubview(m_pTableView);
    
    m_pTableView->onCellAtIndexPath(CALLBACK_BIND_3(MenuViewController::tableCellAtIndex, this));
    m_pTableView->onCellHeightAtIndexPath(CALLBACK_BIND_2(MenuViewController::tableViewHeightForRowAtIndexPath, this));
    m_pTableView->onNumberOfRowsAtIndexPath(CALLBACK_BIND_1(MenuViewController::numberOfRowsInSection, this));
    m_pTableView->onDidSelectCellAtIndexPath(CALLBACK_BIND_2(MenuViewController::tableViewDidSelectRowAtIndexPath, this));
    
    m_pLogo = CAImageView::createWithImage(CAImage::create("image/logo.png"));
    m_pLogo->setLayout(logoLayout);
    this->getView()->addSubview(m_pLogo);
    
}

void MenuViewController::viewDidUnload()
{

}

void MenuViewController::changeStatusBarOrientation(CAObject* obj)
{
    const CAInterfaceOrientation& orientation = CAApplication::getApplication()->getStatusBarOrientation();
    
    DLayout tableViewLayout;
    DLayout logoLayout;
    if (orientation == CAInterfaceOrientation::Landscape)
    {
        tableViewLayout = DLayout(DHorizontalLayoutFill, DVerticalLayout_B_H(0, 400));
        logoLayout = DLayout(DHorizontalLayout_W_C(261, 0.5), DVerticalLayout_T_H(120, 258));
    }
    else
    {
        tableViewLayout = DLayout(DHorizontalLayoutFill, DVerticalLayout_T_B(450, 0));
        logoLayout = DLayout(DHorizontalLayout_W_C(261, 0.5), DVerticalLayout_T_H(120, 258));
    }
    
    m_pTableView->setLayout(tableViewLayout);
    m_pLogo->setLayout(logoLayout);
}

void MenuViewController::tableViewDidSelectRowAtIndexPath(unsigned int section, unsigned int row)
{
    if(row == 0)
    {
        MyWebViewController* webViewController = new MyWebViewController("http://crossapp.9miao.com");
        webViewController->setTitle("CrossApp");
        webViewController->init();
        webViewController->autorelease();
        RootWindow::getInstance()->getRootNavigationController()->pushViewController(webViewController, true);
    }
    else if(row == 1)
    {
        MyWebViewController* webViewController = new MyWebViewController("http://www.9miao.com/");
        webViewController->setTitle("9秒");
        webViewController->init();
        webViewController->autorelease();
        RootWindow::getInstance()->getRootNavigationController()->pushViewController(webViewController, true);
    }
    RootWindow::getInstance()->getDrawerController()->hideLeftViewController(false);
}

CATableViewCell* MenuViewController::tableCellAtIndex(const DSize& cellSize, unsigned int section, unsigned int row)
{
    CATableViewCell* cell = m_pTableView->dequeueReusableCellWithIdentifier("CrossApp");
    if (cell == NULL)
    {
        cell = CATableViewCell::create("CrossApp");
        cell->setBackgroundView(NULL);
        CALabel* test = CALabel::createWithLayout(DLayout(DHorizontalLayout_L_R(50, 0), DVerticalLayoutFill));
        test->setTextAlignment(CATextAlignment::Left);
        test->setVerticalTextAlignmet(CAVerticalTextAlignment::Center);
        test->setFontSize(32);
		test->setColor(CAColor4B::WHITE);
        test->setTag(100);
        cell->addSubview(test);
        
        CAImageView* arrow = CAImageView::createWithLayout(DLayout(DHorizontalLayout_R_W(0, 64), DVerticalLayout_T_H(20, 64)));
        arrow->setTag(101);
        cell->addSubview(arrow);
    }
	CALabel* test = (CALabel*)cell->getSubviewByTag(100);
	test->setText(unicode_to_utf8(menuList[row]));
    CAImageView* arrow = (CAImageView*)cell->getSubviewByTag(101);
    arrow->setImage(CAImage::create("source_material/cell_btn_right.png"));

    return cell;
}

unsigned int MenuViewController::numberOfRowsInSection(unsigned int section)
{
    return 2;
}

unsigned int MenuViewController::tableViewHeightForRowAtIndexPath(unsigned int section, unsigned int row)
{
    return 100;
}

File: samples/Test_JS/Resources/js/MenuViewController.js
/**
 * Created by crossApp on 16/8/8.
 */


var MenuViewController = ca.CAViewController.extend({

    ctor: function () {
        this._super();
    },

    viewDidLoad: function() {

        this.getView().setColor(ca.CAColor4B.CLEAR);

        this.menuList = [];
        this.menuList.push("CrossApp");
        this.menuList.push("9秒");

        var tableViewLayout = ca.DLayout.set(ca.DHorizontalLayout.fill(), ca.DVerticalLayout_T_B(450, 0));
        this.tableView = ca.CATableView.createWithLayout(tableViewLayout);
        this.tableView.setAllowsSelection(true);
        this.tableView.setBackgroundImage(null);
        this.tableView.setSeparatorColor(ca.CAColor4B.set(166, 166, 166, 100));
        this.tableView.setShowsScrollIndicators(false);
        this.tableView.setScrollEnabled(false);
        this.getView().addSubview(this.tableView);

        this.tableView.onCellAtIndexPath(this.tableCellAtIndex.bind(this)) ;
        this.tableView.onCellHeightAtIndexPath(this.tableViewHeightForRowAtIndexPath.bind(this));
        this.tableView.onNumberOfRowsAtIndexPath(this.numberOfRowsInSection.bind(this));
        this.tableView.onDidSelectCellAtIndexPath(this.tableViewDidSelectRowAtIndexPath.bind(this)) ;

        var logoLayout = ca.DLayout.set(ca.DHorizontalLayout_W_C(261, 0.5), ca.DVerticalLayout_T_H(120, 258));
        var logo = ca.CAImageView.createWithImage(ca.CAImage.create("image/logo.png"));
        logo.setLayout(logoLayout);
        this.getView().addSubview(logo);
    },
    tableViewDidSelectRowAtIndexPath: function(section ,row)
    {
        if(row == 0)
        {
            var webViewController = new MyWebViewController("http://crossapp.9miao.com");
            webViewController.setTitle("CrossApp");
            webViewController.init();
            ca.rootWindow.getRootNavigationController().pushViewController(webViewController, true);
        }
        else if(row == 1)
        {
            var webViewController = new MyWebViewController("http://www.9miao.com/");
            webViewController.setTitle("9秒");
            webViewController.init();
            ca.rootWindow.getRootNavigationController().pushViewController(webViewController, true);
        }
        ca.rootWindow.getDrawerController().hideLeftViewController(false);

    },
    tableCellAtIndex: function( cellSize, section ,row)
    {
        var cell = this.tableView.dequeueReusableCellWithIdentifier("CrossApp");
        if (cell == null)
        {
            cell = ca.CATableViewCell.create("CrossApp");
            cell.setBackgroundImage(null, false);
            var test_layout = ca.DLayout.set(ca.DHorizontalLayout_L_R(50, 0),ca.DVerticalLayout.fill());
            var test = ca.CALabel.createWithLayout(test_layout);
            test.setTextAlignment(ca.CATextAlignment.Left);
            test.setVerticalTextAlignmet(ca.CAVerticalTextAlignment.Center);
            test.setFontSize(32);
            test.setColor(ca.CAColor4B.WHITE);
            test.setTag(100);
            cell.addSubview(test);

            var arrow_layout = ca.DLayout.set(ca.DHorizontalLayout_R_W(0, 64),ca.DVerticalLayout_T_H(20, 64));
            var arrow = ca.CAImageView.createWithLayout(arrow_layout);
            arrow.setTag(101);
            cell.addSubview(arrow);
        }
        var test = cell.getSubviewByTag(100);
        test.setText(this.menuList[row]);

        var arrow = cell.getSubviewByTag(101);
        arrow.setImage(ca.CAImage.create("source_material/cell_btn_right.png"));

        return cell;

    },
    numberOfRowsInSection: function(section)
    {
        return 2;
    },
    tableViewHeightForRowAtIndexPath: function(section, row)
    {
        return 100;
    }
});
